### PR TITLE
CB-3725 Port back 'Periscope should restore the instance state automatically' to 2.9

### PR DESCRIPTION
**Pull request by foldik**
Periscope detects new healthy and new failed nodes.

### CODE REVIEW

File: autoscale/src/main/java/com/sequenceiq/periscope/monitor/evaluator/AmbariAgentHealthEvaluator.java [hunk 1]
```diff
@@ -97,12 +107,18 @@ public void execute() {
...
-                    failureReport.setFailedNodes(hostNamesToRecover);
-                    cbClient.autoscaleEndpoint().failureReport(cluster.getStackId(), failureReport);
+                List<FailedNode> failedNodes = failedNodeRepository.findByClusterId(clusterId);
+                Optional<ChangedNodesReport> changedNodesRequest = getChangedNodes(Set.copyOf(hostNamesToRecover), failedNodes);
+                if (changedNodesRequest.isPresent()) {
+                    LOGGER.info("Nodes state changed for cluster {}. New failed nodes {}, new healthy nodes {}.",
+                            clusterId,
+                            changedNodesRequest.get().getNewFailedNodes(),
+                            changedNodesRequest.get().getNewHealthyNodes());
+                    cloudbreakClientConfiguration.cloudbreakClient()
```
Comment (lnardai) [inline autoscale/src/main/java/com/sequenceiq/periscope/monitor/evaluator/AmbariAgentHealthEvaluator.java]: Make sure if periscope can't report the failed nodes to cloudbreak, db won't change it's state.

File: core/src/main/java/com/sequenceiq/cloudbreak/service/cluster/ClusterService.java [hunk 2]
```diff
@@ -433,14 +434,18 @@ public void updateUserNamePassword(Long stackId, UserNamePasswordJson userNamePa
         }
     }
 
-    public void failureReport(Long stackId, List<String> failedNodes) {
+    public void reportHealthChange(Long stackId, Set<String> failedNodes, Set<String> newHealthyNodes) {
+        if (!Sets.intersection(failedNodes, newHealthyNodes).isEmpty()) {
```
Comment (lnardai) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/cluster/ClusterService.java]: Maybe this check can be done in periscope?

Comment (foldik) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/cluster/ClusterService.java]: Others can call this CB API so we need to do it here anyway.

File: autoscale/src/main/java/com/sequenceiq/periscope/monitor/evaluator/AmbariAgentHealthEvaluator.java (hunk 1 above)
Comment (foldik) [reply autoscale/src/main/java/com/sequenceiq/periscope/monitor/evaluator/AmbariAgentHealthEvaluator.java]: Okay, thanks :)

**Pull request merged**